src/platform/backends/hyperv_api/hcs_virtual_machine.cpp

```diff
@@ -0,0 +1,748 @@
+        }
+        case hcs::ComputeSystemState::created:
+            [[fallthrough]];
+        default:
+        {
+            mpl::debug(log_category,
+                       "start() -> VM `{}` is in {} state, starting",
+                       get_name(),
+                       state);
+            return HCS().start_compute_system(hcs_system);
+        }
+        }
+    }();
+
+    if (!status)
+    {
+        state = prev_state;
+        handle_state_update();
+        throw StartComputeSystemException{"Could not start the VM: {}", status};
+    }
+
+    mpl::debug(log_category, "start() -> Start/resume VM `{}`, result `{}`", get_name(), status);
+}
+void HCSVirtualMachine::shutdown(ShutdownPolicy shutdown_policy)
+{
+    mpl::debug(log_category,
+               "shutdown() -> Shutting down VM `{}`, current state {}",
+               get_name(),
+               state);
+
+    switch (shutdown_policy)
+    {
+    case ShutdownPolicy::Powerdown:
+        mpl::debug(log_category,
+                   "shutdown() -> Requested powerdown, initiating graceful shutdown for `{}`",
+                   get_name());
+
+        // If the guest has integration modules enabled, we can use graceful shutdown.
+        if (!HCS().shutdown_compute_system(hcs_system))
+        {
+            // Fall back to SSH shutdown.
+            ssh_exec("sudo shutdown -h now");
+            drop_ssh_session();
+        }
+        break;
+    case ShutdownPolicy::Halt:
+    case ShutdownPolicy::Poweroff:
+        mpl::debug(log_category,
+                   "shutdown() -> Requested halt/poweroff, initiating forceful shutdown for `{}`",
+                   get_name());
+        // These are non-graceful variants. Just terminate the system immediately.
+        const auto r = HCS().terminate_compute_system(hcs_system);
+        mpl::debug(log_category, "shutdown -> terminate_compute_system result: {}", r.code);
+        drop_ssh_session();
+        break;
+    }
+
+    // We need to wait here.
+    auto on_timeout = [] {
+        throw std::runtime_error("timed out waiting for initialization to complete");
+    };
+
+    multipass::utils::try_action_for(on_timeout, std::chrono::seconds{180}, [this]() {
+        set_state(fetch_state_from_api());
+        switch (current_state())
+        {
+        case VirtualMachine::State::stopped:
+        case VirtualMachine::State::off:
+            return multipass::utils::TimeoutAction::done;
+        default:
+            return multipass::utils::TimeoutAction::retry;
+        }
+    });
+}
+
+void HCSVirtualMachine::suspend()
+{
+    mpl::debug(log_category,
+               "suspend() -> Suspending VM `{}`, current state {}",
+               get_name(),
+               state);
+    const auto& [status, status_msg] = HCS().pause_compute_system(hcs_system);
+    set_state(fetch_state_from_api());
+    handle_state_update();
+}
+
+HCSVirtualMachine::State HCSVirtualMachine::current_state()
+{
+    return state;
```

Comment:
On managed backends, this normally updates the state first. That's also the case in the old Hyper-V. Why the different behavior here?